### PR TITLE
[WIP] feat: improve error message when Table methods shadow column names

### LINKED ISSUE #5062: ux: count aggregates create column named "count" which isn't super friendly
### What happened?

All of our count aggregates create a column named "count" in the outputs.  Consider the silly example below:

```python
import pandas as pd
import ibis

[ins] In [11]: df = pd.DataFrame({"x": [1.0, 2.0, 3.0, 1.0], "y": ["a", "b", "c", "a"]})

[ins] In [12]: t = ibis.memtable(df)

[ins] In [13]: t
Out[13]: 
┏━━━━━━━━━┳━━━━━━━━┓
┃ x       ┃ y      ┃
┡━━━━━━━━━╇━━━━━━━━┩
│ float64 │ string │
├─────────┼────────┤
│     1.0 │ a      │
│     2.0 │ b      │
│     3.0 │ c      │
│     1.0 │ a      │
└─────────┴────────┘


[ins] In [14]: t.x.value_counts().join(t, "x")
Out[14]: 
┏━━━━━━━━━┳━━━━━━━┳━━━━━━━━┓
┃ x       ┃ count ┃ y      ┃
┡━━━━━━━━━╇━━━━━━━╇━━━━━━━━┩
│ float64 │ int64 │ string │
├─────────┼───────┼────────┤
│     1.0 │     2 │ a      │
│     2.0 │     1 │ b      │
│     3.0 │     1 │ c      │
│     1.0 │     2 │ a      │
└─────────┴───────┴────────┘


[ins] In [15]: t.x.value_counts().join(t, "x").order_by(ibis._.count)
```

The last line fails because `count` is very much a method on tables.

There are other ways to spell this -- either of

```
t.x.value_counts().join(t, "x").order_by("count")
t.x.value_counts().join(t, "x").order_by(ibis._.["count"])
```

Naming collisions happen, but this one is a little bit aggravating because it's a column name that Ibis creates that conflicts with itself.

@jcrist suggested that we might be able to do some descriptor stuff to at least raise a better error message.  
Altenatively, we could pick a different name for that column (although I admit I'm not sure what that would be).

This may be intractable, but I wanted to raise it.

### What version of ibis are you using?

master

### What backend(s) are you using, if any?

all of them, all the time

### Relevant log output

```sh
---------------------------------------------------------------------------
IbisTypeError                             Traceback (most recent call last)
Cell In[15], line 1
----> 1 t.x.value_counts().join(t, "x").order_by(ibis._.count)

File ~/github.com/ibis-project/ibis/ibis/expr/types/relations.py:477, in Table.order_by(self, by)
    475 else:
    476     by = util.promote_list(by)
--> 477 return self.op().order_by(by).to_expr()

File ~/github.com/ibis-project/ibis/ibis/expr/operations/relations.py:366, in Selection.order_by(self, sort_exprs)
    363 def order_by(self, sort_exprs):
    364     from ibis.expr.analysis import shares_all_roots, sub_immediate_parents
--> 366     keys = rlz.tuple_of(rlz.sort_key_from(rlz.just(self)), sort_exprs)
    368     if not self.selections:
    369         if shares_all_roots(keys, table := self.table):

File /nix/store/crr9ya0n4hnnay9myvffdrlzpkrcrws3-python3-3.10.9-env/lib/python3.10/site-packages/toolz/functoolz.py:304, in curry.__call__(self, *args, **kwargs)
    302 def __call__(self, *args, **kwargs):
    303     try:
--> 304         return self._partial(*args, **kwargs)
    305     except TypeError as exc:
    306         if self._should_curry(args, kwargs, exc):

File ~/github.com/ibis-project/ibis/ibis/common/validators.py:194, in container_of(inner, arg, type, min_length, flatten, **kwargs)
    191 if flatten:
    192     arg = flatten_iterable(arg)
--> 194 return type(inner(item, **kwargs) for item in arg)

File ~/github.com/ibis-project/ibis/ibis/common/validators.py:194, in <genexpr>(.0)
    191 if flatten:
    192     arg = flatten_iterable(arg)
--> 194 return type(inner(item, **kwargs) for item in arg)

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:84, in rule.__call__(self, *args, **kwargs)
     82 args = map(self._erase_expr, args)
     83 kwargs = {k: self._erase_expr(v) for k, v in kwargs.items()}
---> 84 result = super().__call__(*args, **kwargs)
     85 assert not isinstance(result, ir.Expr)
     86 return result

File /nix/store/crr9ya0n4hnnay9myvffdrlzpkrcrws3-python3-3.10.9-env/lib/python3.10/site-packages/toolz/functoolz.py:304, in curry.__call__(self, *args, **kwargs)
    302 def __call__(self, *args, **kwargs):
    303     try:
--> 304         return self._partial(*args, **kwargs)
    305     except TypeError as exc:
    306         if self._should_curry(args, kwargs, exc):

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:114, in sort_key_from(table_ref, key, **kwargs)
    102 is_ascending = {
    103     "asc": True,
    104     "ascending": True,
   (...)
    110     True: True,
    111 }
    113 if callable(key):
--> 114     key = function_of(table_ref, key, **kwargs)
    116 if isinstance(key, ops.SortKey):
    117     return key

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:84, in rule.__call__(self, *args, **kwargs)
     82 args = map(self._erase_expr, args)
     83 kwargs = {k: self._erase_expr(v) for k, v in kwargs.items()}
---> 84 result = super().__call__(*args, **kwargs)
     85 assert not isinstance(result, ir.Expr)
     86 return result

File /nix/store/crr9ya0n4hnnay9myvffdrlzpkrcrws3-python3-3.10.9-env/lib/python3.10/site-packages/toolz/functoolz.py:304, in curry.__call__(self, *args, **kwargs)
    302 def __call__(self, *args, **kwargs):
    303     try:
--> 304         return self._partial(*args, **kwargs)
    305     except TypeError as exc:
    306         if self._should_curry(args, kwargs, exc):

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:512, in function_of(table_ref, fn, output_rule, this)
    507 else:
    508     raise com.IbisTypeError(
    509         'argument `fn` must be a function, lambda or deferred operation'
    510     )
--> 512 return output_rule(arg, this=this)

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:84, in rule.__call__(self, *args, **kwargs)
     82 args = map(self._erase_expr, args)
     83 kwargs = {k: self._erase_expr(v) for k, v in kwargs.items()}
---> 84 result = super().__call__(*args, **kwargs)
     85 assert not isinstance(result, ir.Expr)
     86 return result

File /nix/store/crr9ya0n4hnnay9myvffdrlzpkrcrws3-python3-3.10.9-env/lib/python3.10/site-packages/toolz/functoolz.py:304, in curry.__call__(self, *args, **kwargs)
    302 def __call__(self, *args, **kwargs):
    303     try:
--> 304         return self._partial(*args, **kwargs)
    305     except TypeError as exc:
    306         if self._should_curry(args, kwargs, exc):

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:210, in value(dtype, arg, **kwargs)
    206 import ibis.expr.operations as ops
    208 if not isinstance(arg, ops.Value):
    209     # coerce python literal to ibis literal
--> 210     arg = literal(None, arg)
    212 if dtype is None:
    213     # no datatype restriction
    214     return arg

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:84, in rule.__call__(self, *args, **kwargs)
     82 args = map(self._erase_expr, args)
     83 kwargs = {k: self._erase_expr(v) for k, v in kwargs.items()}
---> 84 result = super().__call__(*args, **kwargs)
     85 assert not isinstance(result, ir.Expr)
     86 return result

File /nix/store/crr9ya0n4hnnay9myvffdrlzpkrcrws3-python3-3.10.9-env/lib/python3.10/site-packages/toolz/functoolz.py:304, in curry.__call__(self, *args, **kwargs)
    302 def __call__(self, *args, **kwargs):
    303     try:
--> 304         return self._partial(*args, **kwargs)
    305     except TypeError as exc:
    306         if self._should_curry(args, kwargs, exc):

File ~/github.com/ibis-project/ibis/ibis/expr/rules.py:178, in literal(dtype, value, **kwargs)
    176     dtype = inferred_dtype
    177 else:
--> 178     raise com.IbisTypeError(
    179         'The datatype of value {!r} cannot be inferred, try '
    180         'passing it explicitly with the `type` keyword.'.format(value)
    181     )
    183 if dtype.is_null():
    184     return ops.NullLiteral()

IbisTypeError: The datatype of value <bound method Table.count of ┏━━━━━━━━━┳━━━━━━━┳━━━━━━━━┓
┃ x       ┃ count ┃ y      ┃
┡━━━━━━━━━╇━━━━━━━╇━━━━━━━━┩
│ float64 │ int64 │ string │
├─────────┼───────┼────────┤
│     1.0 │     2 │ a      │
│     2.0 │     1 │ b      │
│     3.0 │     1 │ c      │
│     1.0 │     2 │ a      │
└─────────┴───────┴────────┘
> cannot be inferred, try passing it explicitly with the `type` keyword.
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jcrist**
This adds a method decorator to be used on `Table` methods. The intent is to provide nicer error messages when the user tries to use attribute-access to get a column, but a method on the table shadows the column. Previously you'd get a cryptic error, we now try to provide a nicer error message.

For now this is only applied to `Table.count`, but could be used elsewhere. Just putting this up for discussion for now.

```python
In [1]: import ibis                                                                           
                                                                                              
In [2]: from ibis import _                                                                    
                                                                                              
In [3]: t = ibis.table({"x": "int", "count": "int"})                                          
                                                                                              
In [4]: t.filter(_.count > 10)
---------------------------------------------------------------------------                                                                                                                  
TypeError                                 Traceback (most recent call last)                                                                                                                  
Input In [4], in <cell line: 1>()                                                                                                                                                            
----> 1 t.filter(_.count > 10)                                                                                                                                                               
...
TypeError: '>' not supported for method `Table.count`. If you meant to access a column named 'count' use `table['count']` syntax instead.
```

Fixes #5062.

### CODE REVIEW

**Comment by kszucs** [comment]
This is a nice ergonomical addition, but requires some weird trickery. Though I don't like this trickery, I cannot think any alternative implementation to achieve this.

File: ibis/expr/table_method.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import functools
```
Comment (kszucs) [inline ibis/expr/table_method.py:1]: Could you please move the content of this file to `ibis.util`?

File: ibis/expr/table_method.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import inspect
+
+
+class bound_table_method:
```
Comment (kszucs) [inline ibis/expr/table_method.py:5]: Can we subclass `types.MethodType` se we return with a valid method?

File: ibis/expr/table_method.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    @property
+    def __doc__(self):
+        return self._func.__doc__
+
+    @property
+    def __name__(self):
+        return self._func.__name__
+
+    @property
+    def __signature__(self):
```
Comment (kszucs) [inline ibis/expr/table_method.py:23]: Why is this required?

File: ibis/expr/table_method.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        raise TypeError(
+            self._errmsg("{sym!r} not supported for method `Table.{name}`.", sym=sym)
+        )
+
+    __eq__ = functools.partialmethod(_compare, sym="=")
+    __ne__ = functools.partialmethod(_compare, sym="!=")
+    __lt__ = functools.partialmethod(_compare, sym="<")
+    __le__ = functools.partialmethod(_compare, sym="<=")
+    __gt__ = functools.partialmethod(_compare, sym=">")
+    __ge__ = functools.partialmethod(_compare, sym=">=")
```
Comment (kszucs) [inline ibis/expr/table_method.py:58]: Could there be other operators which require overloads?

File: ibis/expr/table_method.py (hunk 2 above)
Comment (kszucs) [inline ibis/expr/table_method.py:5]: This looks like a mixin of a more generic proxy type (like [this one](https://github.com/PEAK-Legacy/ProxyTypes#proxy-basics)) and a method. 
We already have deferred which also acts like a proxy type so it could make sense to introduce a generic proxy class and use that both here and for deferred.

File: ibis/expr/table_method.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class table_method:
```
Comment (kszucs) [inline ibis/expr/table_method.py:61]: Is this required only to support `t.count.min()`?

File: ibis/expr/table_method.py (hunk 2 above)
Comment (jcrist) [reply ibis/expr/table_method.py:5]: `types.MethodType` isn't a subclassable type. Most of the complications here are around mocking out a descriptor that functions like a `MethodType`  but lets us provide nicer error messages.

File: ibis/expr/table_method.py (hunk 1 above)
Comment (jcrist) [reply ibis/expr/table_method.py:1]: Since this is tightly coupled with the expression system (for generating errors), I figured here was a better fit. It's not a general utility.

File: ibis/expr/table_method.py (hunk 3 above)
Comment (jcrist) [reply ibis/expr/table_method.py:23]: For signature inference of the method. `__signature__` is used by `inspect.signature` and other introspection tools (this mainly comes up when working in `ipython`/`jupyter`).

File: ibis/expr/table_method.py (hunk 4 above)
Comment (jcrist) [reply ibis/expr/table_method.py:58]: Yeah, probably any binary operator honestly. This is mainly to catch things like `t.count > 1`, but `t.count + 1` could just as likely happen.

**Comment by jcrist** [comment]
After some thought, I'm not sure if the complications here are worth it. I'm happy to push through with this PR, but I'm just as happy to close it - it's a lot of magic just to provide some better error messages in certain situations.

**Comment by jcrist** [comment]
Yeah, nixing this as too much work/magic. Closing this for now, can always reuse the work later if we decide it's worth it.
